Homework 5.cpp

```diff
@@ -0,0 +1,153 @@
+//
+//  Homework 5.cpp
+//  StudyFall16
+//
+//  Created by Farida Eid on 7/8/17.
+//  Copyright © 2017 Farida Eid. All rights reserved.
+//
+
+#include <iostream>
+#include <vector>
+#include <map>
+#include <unordered_map>
+#include <stack>
+#include <set>
+
+using namespace std;
+
+class Graph
+{
+private:
+    map<char,set<char>> adjList;
+    unordered_map<char, bool> visited;
+    void topologicalSortRecursive(char, stack<char>&);
+    
+public:
+    Graph();
+    void addEdge(char, char);
+    void topologicalSort(vector<char>&);
+    
+};
+
+// Graph constructor
+// Parameters: No parameters
+Graph::Graph()
+{
+}
+
+//Function to add edge in the graph
+//Parameters: Vertix 1 and Vertix 2
+void Graph::addEdge(char u, char v)
+{
+    visited[u] = false;
+    visited[v] = false;
+    
+    adjList[u].insert(v);
+}
+
+
+
+// Function to perform topological sort
+// Parameters: The vector of characters to save the answer in
+void Graph::topologicalSort(vector<char>& characters)
+{
+    stack<char> Stack;
+    
+    map<char, set<char>>::iterator it;
+    
+    
+    for(it = adjList.begin(); it != adjList.end(); it++)
+    {
+        if(!visited[it->first])
+            topologicalSortRecursive(it->first, Stack);
+    }
+    
+    while(!Stack.empty())
+    {
+        characters.push_back(Stack.top());
+        
+        Stack.pop();
+    }
+    
+}
+
+// Recursive function to complete the topological sort
+// Paramters: Current Index, visited array to mark the vertices, Stack to add the answer
+void Graph::topologicalSortRecursive(char current , stack<char> & Stack)
+{
+    visited[current] = true;
+    
+    set<char>::iterator it;
+
+    for(it = adjList[current].begin(); it != adjList[current].end(); it++)
+    {
+        if(!visited[(*it)])
+            topologicalSortRecursive((*it), Stack);
+        
+    }
+    
+    Stack.push(current);
+}
+
+// Function that takes a list of words, create a graph and perform topological sort
+// Parameters: list of words, list of characters to save the answer in
+vector<char> findAlphabet(const vector<string>& unknownLang)
+{
+    Graph graph;
+    
+    string word1, word2;
+    int length;
+    vector<char> characters;
+    
+    for(int index = 0; index < unknownLang.size() -1; index++)
```

Comment:
There is a very subtle bug here, unknownLang.size() is unsigned, so when it's 0 and you subtract one you end up with a very large number, not -1. Things don't go well after that :)